File: app/src/main/java/com/tablehockey/game/game/Puck.kt
package com.tablehockey.game.game

import kotlin.math.hypot

class Puck {
    var x = 0f
    var y = 0f
    var vx = 0f
    var vy = 0f

    /** Skater currently carrying the puck on their stick, or null when loose. */
    var carrier: Skater? = null

    /** True while the puck is travelling as a shot (until it is touched again). */
    var shot = false
    var shooter: Skater? = null
    var passTarget: Skater? = null
    var lastTouchTeam = -1
    /** True while a shot is slipping through the goalie (difficulty leak); cleared once clear of the pads. */
    var leaking = false

    // Network interpolation targets (client only)
    var netX = 0f
    var netY = 0f

    val speed: Float get() = hypot(vx, vy)

    fun reset(px: Float, py: Float) {
        x = px; y = py; vx = 0f; vy = 0f
        carrier = null
        shot = false
        shooter = null
        passTarget = null
        leaking = false
        netX = px; netY = py
    }
}

File: ios/PowerPlayHockeyCore/Sources/PowerPlayHockeyCore/Game/Puck.swift
import Foundation

final class Puck {
    var x: Float = 0
    var y: Float = 0
    var vx: Float = 0
    var vy: Float = 0

    /// Skater currently carrying the puck on their stick, or nil when loose.
    var carrier: Skater?

    /// True while the puck is travelling as a shot (until it is touched again).
    var shot: Bool = false
    var shooter: Skater?
    var passTarget: Skater?
    var lastTouchTeam: Int = -1
    /// True while a shot is slipping through the goalie (difficulty leak);
    /// cleared once clear of the pads.
    var leaking: Bool = false

    // Network interpolation targets (client only)
    var netX: Float = 0
    var netY: Float = 0

    var speed: Float { hypot(vx, vy) }

    func reset(_ px: Float, _ py: Float) {
        x = px; y = py; vx = 0; vy = 0
        carrier = nil
        shot = false
        shooter = nil
        passTarget = nil
        leaking = false
        netX = px; netY = py
    }
}
